dds/DCPS/security/AccessControl/XmlUtils.cpp

```diff
@@ -112,48 +114,268 @@ bool parse_bool(const XMLCh* in, bool& value)
+   *   means. The DDS Security 1.1 spec explicitly says in 9.4.1.3.2.2 that it
+   *   would be UTC. Xerces getEpoch values match up with this in the unit test
+   *   for this code, so it's also interpreting them as UTC instead of
+   *   something like the local system timezone.
+   * - Since we use time_t we will accept but ignore fractional seconds.
+   *   NOTE: The security spec is missing fractional seconds in a format
+   *   description in a comment in example XML. This comment has been copied
+   *   into many permissions files sitting in OpenDDS.
+   */
+
+  const std::string str = to_string(in);
+  if (str[0] == '-') {
+    if (security_debug.access_error) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: {access_error} parse_time: "
+        "date/time can't be negative: \"%C\"\n",
+        str.c_str()));
+    }
+    return false;
+  }
+
+  /*
+   * After this the expected lexical format given by the Schema spec is:
+   *   yyyy '-' mm '-' dd 'T' hh ':' mm ':' ss ('.' s+)? ((('+' | '-') hh ':' mm) | 'Z')?
+   * or alternatively as the Security spec says, except with fractional seconds:
+   *   CCYY-MM-DDThh:mm:ss[.fffff...][Z|(+|-)hh:mm]
+   * See unit tests for this code for examples
+   */
+  std::tm dttm;
+  size_t pos = 0;
+
+  // Year
+  if (!parse_time_field(str, pos, dttm.tm_year, 4, -1900)) {
+    return false;
+  }
+  if (!parse_time_char(str, pos, "-")) {
+    return false;
+  }
+  // Month
+  if (!parse_time_field(str, pos, dttm.tm_mon, 2, -1)) {
+    return false;
+  }
+  if (!parse_time_char(str, pos, "-")) {
+    return false;
+  }
+  // Day
+  if (!parse_time_field(str, pos, dttm.tm_mday)) {
+    return false;
+  }
+  if (!parse_time_char(str, pos, "T")) {
+    return false;
+  }
+  // Hour
+  if (!parse_time_field(str, pos, dttm.tm_hour)) {
+    return false;
+  }
+  if (!parse_time_char(str, pos, ":")) {
+    return false;
+  }
+  // Minutes
+  if (!parse_time_field(str, pos, dttm.tm_min)) {
+    return false;
+  }
+  if (!parse_time_char(str, pos, ":")) {
+    return false;
+  }
+  // Seconds
+  if (!parse_time_field(str, pos, dttm.tm_sec)) {
+    return false;
+  }
+
+  // Ignore Fractional Seconds
+  if (str[pos] == '.') {
+    ++pos;
+    for (; str[pos] >= '0' && str[pos] <= '9'; ++pos) {
+    }
+  }
+
+  // Optional Timezone Info
+  time_t timezone_offset = 0;
+  char tz_char;
+  bool end;
+  if (!parse_time_char_or_end(str, pos, "+-Z", tz_char, end)) {
+    return false;
+  }
+  if (!end && tz_char != 'Z') {
+    int tz_hour;
+    if (!parse_time_field(str, pos, tz_hour)) {
+      return false;
+    }
+
+    if (!parse_time_char(str, pos, ":")) {
+      return false;
+    }
+
+    int tz_min;
+    if (!parse_time_field(str, pos, tz_min)) {
+      return false;
+    }
+
+    /*
+     * We will have to do the reverse of the sign to convert the time timezone
+     * to the UTC. For example 00:00:00+00:00 (Midnight UTC) is 06:00:00-06:00
+     * (6 AM CST).
```

Comment:
CST is indeed UTC-06:00, 6 hours behind UTC.  So midnight in UTC would be 6 PM in CST.

Reply:
Thanks, I mixed that up.